[Cocoa] Default subtitle font is wrong (serif vs. sans-serif)

jernoble opened the pull request:
#### 414c1f8ec90f29e67980711d76f99522ddd5ea3a
<pre>
[Cocoa] Default subtitle font is wrong (serif vs. sans-serif)
<a href="https://bugs.webkit.org/show_bug.cgi?id=276449">https://bugs.webkit.org/show_bug.cgi?id=276449</a>
<a href="https://rdar.apple.com/126094074">rdar://126094074</a>

Reviewed by Eric Carlson.

MediaAccessibility will vend a CTFontDescription whose font name is non-public (starting with a &quot;.&quot;). In this case,
convert the font name to either &quot;system-ui&quot; or &quot;system-ui-monospaced&quot; depending on the non-public font name.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsDefaultFontCSS const):

Canonical link: <a href="https://commits.webkit.org/280864@main">https://commits.webkit.org/280864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/512193ecd9c779ed37dd0b41e2241b0a836247e1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57794 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37122 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61416 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59922 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44758 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8427 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46825 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5845 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59824 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34817 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49951 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27654 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31586 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7244 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7243 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53538 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7514 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63099 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1708 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7582 "Found 1 new test failure: media/video-playsinline.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54047 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1714 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49962 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54169 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12801 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1451 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32951 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34037 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35121 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33782 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
```diff
@@ -461,6 +462,18 @@ String CaptionUserPreferencesMediaAF::captionsDefaultFontCSS() const
+    if (fontNameIsSystemFont(name.get())) {
+        if (CFStringHasPrefix(CFSTR(".AppleSystemUIFontMonospaced"), name.get()))
+            name = CFSTR("system-ui-monospaced");
+        else if (CFStringHasPrefix(CFSTR(".AppleSystemUIFont"), name.get()))
```
The `else if` and `else` produce the same result.
Do we need to keep the `else if`?

jernoble replied:
I'm setting us up to add more `else if` cases without a bunch of re-writing existing conditions.

fantasai commented on Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
```diff
@@ -461,6 +462,18 @@ String CaptionUserPreferencesMediaAF::captionsDefaultFontCSS() const
 
+    if (fontNameIsSystemFont(name.get())) {
+        if (CFStringHasPrefix(CFSTR(".AppleSystemUIFontMonospaced"), name.get()))
+            name = CFSTR("system-ui-monospaced");
```
This should be `ui-monospace` per https://www.w3.org/TR/css-fonts-4/#valdef-font-family-ui-monospace

and since that isn't guaranteed to exist on a system, I'd add a fallback to the list (either `system-ui` or `monospace`, depending on what you want to prioritize).

jernoble committed:
[Cocoa] Default subtitle font is wrong (serif vs. sans-serif)
https://bugs.webkit.org/show_bug.cgi?id=276449
rdar://126094074

Reviewed by Eric Carlson.

MediaAccessibility will vend a CTFontDescription whose font name is non-public (starting with a "."). In this case,
convert the font name to either "system-ui" or "system-ui-monospaced" depending on the non-public font name.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsDefaultFontCSS const):

Canonical link: https://commits.webkit.org/280864@main
